File: src/ru/hudspeed/pro/MapOrientation.java
package ru.gpsantiradar.app;

/** Pure heading decisions shared by location-marker and map-camera updates. */
public final class MapOrientation {
    private static final float MOVING_SPEED_KMH = 3f;

    private MapOrientation() {}

    public static float stableHeading(float previousHeading, float measuredHeading,
                                      float speedKmh) {
        if (speedKmh < MOVING_SPEED_KMH || Float.isNaN(measuredHeading)
                || Float.isInfinite(measuredHeading)) {
            return previousHeading;
        }
        return Geo.normalize(measuredHeading);
    }

    public static float cameraAzimuth(boolean autoRotate, float speedKmh,
                                      float heading, float currentAzimuth) {
        return autoRotate && speedKmh >= MOVING_SPEED_KMH
                ? heading : currentAzimuth;
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/MapOrientation.kt
package ru.gpsantiradar.app

/** Pure heading decisions shared by location-marker and map-camera updates. */
object MapOrientation {
    private const val MOVING_SPEED_KMH = 3f

    fun stableHeading(previousHeading: Float, measuredHeading: Float, speedKmh: Float): Float {
        if (speedKmh < MOVING_SPEED_KMH || measuredHeading.isNaN() || measuredHeading.isInfinite()) return previousHeading
        return Geo.normalize(measuredHeading)
    }

    fun cameraAzimuth(autoRotate: Boolean, speedKmh: Float, heading: Float, currentAzimuth: Float): Float =
        if (autoRotate && speedKmh >= MOVING_SPEED_KMH) heading else currentAzimuth
}
